source/extensions/access_loggers/common/access_log_base.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/config/filter/accesslog/v2/accesslog.pb.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/server/access_log_config.h"
+
+#include "common/http/header_map_impl.h"
+#include "common/http/header_utility.h"
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace Common {
+
+/**
+ * Base implementation of Accesslog::Instance handles common filter logic.
+ */
+class ImplBase : public AccessLog::Instance {
+public:
+  ImplBase(AccessLog::FilterPtr filter) : filter_(std::move(filter)) {}
+
+  /**
+   * Log a completed request if the underlying AccessLog `filter_` allows it.
+   */
+  virtual void log(const Http::HeaderMap* request_headers, const Http::HeaderMap* response_headers,
+                   const Http::HeaderMap* response_trailers,
+                   const StreamInfo::StreamInfo& stream_info) override {
+    static Http::HeaderMapImpl empty_headers;
+    if (!request_headers) {
+      request_headers = &empty_headers;
+    }
+    if (!response_headers) {
+      response_headers = &empty_headers;
+    }
+    if (!response_trailers) {
+      response_trailers = &empty_headers;
+    }
+    if (filter_ &&
+        !filter_->evaluate(stream_info, *request_headers, *response_headers, *response_trailers)) {
+      return;
+    }
+    return emitLog(*request_headers, *response_headers, *response_trailers, stream_info);
+  }
```

Comment:
this seems like too large a function to inline into a .h as a virtual function. Put it into a .cc?